File: cpp/leetcode/384.cpp
class Solution {
    vector<int> arr, idx;
public:
    Solution(vector<int> nums) {
        arr.resize(nums.size());
        idx.resize(nums.size());
        for (int i = 0; i < nums.size(); i++){
            arr[i] = nums[i];
            idx[i] = nums[i];
        }
    }
    
    vector<int> reset() {
        for (int i = 0; i < arr.size(); i++)
            arr[i] = idx[i];
        return arr;    
    }
    
    vector<int> shuffle() {
         for(int i = arr.size() - 1; i > 0; i--) {
            int j = rand() % (i + 1);
            swap(arr[i], arr[j]);
         }
         return arr;    
    }
};

File: python/leetcode/384.py
from random import randint
class Solution(object):
    def __init__(self, nums):
        self.nums = nums

    def reset(self):
        return self.nums

    def shuffle(self):
        res, n = self.nums[:], len(self.nums)
        for i in range(n):
            r = random.randint(0,i) #[0, i]
            res[i], res[r] = res[r], res[i]
        return res

        # idx = shuffle(n)
        # return [self.nums[i] for i in idx]